library/src/main/java/com/novoda/downloadmanager/DownloadBatch.java

```diff
@@ -47,51 +39,49 @@
     }
 
     void setCallback(DownloadBatchStatusCallback callback) {
-        this.callback = callback;
         callbackThrottle.setCallback(callback);
     }
 
     void download() {
         String rawBatchId = downloadBatchStatus.getDownloadBatchId().rawId();
         Logger.v("start sync download " + rawBatchId + ", " + STATUS + " " + downloadBatchStatus.status());
 
-        if (shouldAbortStartingBatch(connectionChecker, callback, downloadBatchStatus, downloadsBatchPersistence)) {
+        if (shouldAbortStartingBatch(connectionChecker, callbackThrottle, downloadBatchStatus, downloadsBatchPersistence)) {
             Logger.v("abort starting download " + rawBatchId + ", " + STATUS + " " + downloadBatchStatus.status());
             return;
         }
 
-        markAsDownloadingIfNeeded(downloadBatchStatus, downloadsBatchPersistence, callback);
+        markAsDownloadingIfNeeded(downloadBatchStatus, downloadsBatchPersistence, callbackThrottle);
 
         updateTotalSize();
 
         Logger.v("batch " + downloadBatchStatus.getDownloadBatchId().rawId()
                          + " " + STATUS + " " + downloadBatchStatus.status()
                          + " totalBatchSize " + totalBatchSizeBytes);
 
-        if (shouldAbortAfterGettingTotalBatchSize(downloadBatchStatus, downloadsBatchPersistence, callback, totalBatchSizeBytes)) {
+        if (shouldAbortAfterGettingTotalBatchSize(downloadBatchStatus, downloadsBatchPersistence, callbackThrottle, totalBatchSizeBytes)) {
             Logger.v("abort after getting total batch size download " + rawBatchId + ", " + STATUS + " " + downloadBatchStatus.status());
             return;
         }
 
         for (DownloadFile downloadFile : downloadFiles) {
-            if (batchCannotContinue(downloadBatchStatus, connectionChecker, downloadsBatchPersistence, callback)) {
+            if (batchCannotContinue(downloadBatchStatus, connectionChecker, downloadsBatchPersistence, callbackThrottle)) {
                 break;
             }
             downloadFile.download(fileDownloadCallback);
         }
 
         if (networkError(downloadBatchStatus)) {
-            processNetworkError(downloadBatchStatus, callback, downloadsBatchPersistence);
+            processNetworkError(downloadBatchStatus, callbackThrottle, downloadsBatchPersistence);
         }
 
-        deleteBatchIfNeeded(downloadBatchStatus, downloadsBatchPersistence, callback);
-        notifyCallback(callback, downloadBatchStatus);
-        callbackThrottle.stopUpdates();
```

Comment:
⚠️ ❓ I don't think this method if required anymore on the `CallbackThrottle`. The `CallbackThrottle` will always be able to callback to the `Downloader` for a given `Batch`. If the batch is removed then it is removed from the map, when it is removed from the map then this throttle will be eligible for GC and will not exist further. Let me know what you think please 😄 

I haven't removed it because this is a discussion before I remove a potentially controversial piece of code.

Reply:
mmm for every single download there is a new callback being created, this also shuts down any call inside the implementation of the callback that should be called when that download finishes, check for example `callbackThrottleByTime.stopUpdates`. I think we have to call it